Mutation table pagination

adamabeshouse opened the pull request.

Commit:
Pagination for enhanced react table, used in mutation table

adamabeshouse commented:
do you know why this is happening?
![image](https://cloud.githubusercontent.com/assets/636232/21920562/f0bb8938-d92f-11e6-8389-4f1d3a37aa18.png)

onursumer commented on src/shared/components/enhancedReactTable/EnhancedReactTable.tsx:
```diff
@@ -148,12 +176,24 @@ export default class EnhancedReactTable extends React.Component<IEnhancedReactTa
                     onColumnToggled={this.handleVisibilityToggle}
                     showSearch={true}
+                    className="pull-right"
+                    paginationProps={{itemsPerPage:this.state.itemsPerPage,
```
Should we create a secondary `TableHeaderControls` at the end of the table for pagination?

adamabeshouse replied:
"TableFooterControls" ?

onursumer replied:
No need to rename the component itself :) We can have multiple instances of it if we think header is getting too crowded. We don't need to worry about it right now I guess.

onursumer commented on src/shared/components/enhancedReactTable/EnhancedReactTable.tsx:
```diff
@@ -165,6 +205,15 @@ export default class EnhancedReactTable extends React.Component<IEnhancedReactTa
+        if (itemsPerPage === -1) {
+            return 1;
+        } else {
+            return Math.ceil(this.props.rawData.length / itemsPerPage);
```
Actually, we need the length of the "filtered data" here, otherwise we end up with a lot of empty pages when the table is filtered

adamabeshouse replied:
done

onursumer commented on src/shared/components/enhancedReactTable/IEnhancedReactTableProps.ts:
```diff
@@ -12,11 +12,16 @@ export interface IEnhancedReactTableProps {
     rawData: Array<any>; // raw data
+    initItemsPerPage?: number; // initial number of items per page
+    initPage?:number; // initial page
+    itemsName?:string; // name of items, e.g. "mutations"
```
`itemsName` doesn't need to be a property of `EnhancedReactTable`, it might be better to pass this in the `headerControlsProps`.

adamabeshouse replied:
I'm not using it anymore anyway

onursumer commented on src/shared/components/mutationTable/MutationTable.tsx:
```diff
@@ -161,6 +161,7 @@ export default class MutationTable extends React.Component<IMutationTableProps,
         return(
             <div>
                 <EnhancedReactTable
+                    itemsName="mutations"
```
move into `headerControlsProps`?

onursumer commented on src/shared/components/enhancedReactTable/EnhancedReactTable.tsx:
```diff
@@ -108,6 +111,28 @@ export default class EnhancedReactTable extends React.Component<IEnhancedReactTa
+        this.onNextPageClick = this.onNextPageClick.bind(this);
+    }
+
+    onChangeItemsPerPage(itemsPerPage:number) {
```
Can we make these methods private, and move them right after (or right before) the other state change handlers for consistency?

adamabeshouse replied:
done

onursumer reviewed (changes requested):
Main issues need to be resolved:
- Dropbox for the page size has a different/default style
- When the table is filtered, we end up with lots of empty pages

Commit:
Changes requested by Onur

mimic bootstrap style for page size select

remove unused itemsName prop and make handler methods private in EnhancedReactTable

set up hacky way to get at filtered data from Table for pagination info

alisman commented on src/pages/patientView/mutation/MutationInformationContainer.tsx:
```diff
@@ -29,6 +29,12 @@ export default class MutationInformationContainer extends React.Component<IMutat
 
+    componentWillUpdate(nextProps:IMutationInformationContainerProps, nextState:{}) {
+        if (nextProps.mutations !== this.props.mutations) {
+            this.mergedMutations = this.mergeMutations(nextProps.mutations);
```
what does this do?

adamabeshouse replied:
into the MutationTable, we pass in an array of arrays, where each subarray is different instances of the same mutation (i.e. on different samples).

mergeMutations groups together the individual mutation events into these arrays so that the same mutation events are grouped together.

What comes in is a flat list of mutations, so whenever the props get changed, this computation needs to be done again.

alisman replied:
see my slack comment about Reselect library, which I think is a more "idiomatic" way to handle the memoization of this derivation

alisman commented on src/shared/components/enhancedReactTable/EnhancedReactTable.tsx:
```diff
@@ -393,4 +468,23 @@ export default class EnhancedReactTable<T> extends React.Component<IEnhancedReac
     }
+
+    private handleChangeItemsPerPage(itemsPerPage:number) {
+        const nextState = Object.assign({}, this.state);
```
is this to get around ts issue with partial setState?  if so, i would rather we use the TS hack for this.  casting the object as the type of the state.

adamabeshouse replied:
come to think of it I don't understand why this succeeds in typescript...it can't statically check this. This is just as deceptive to TS as casting. I'll change to casting.

adamabeshouse replied:
done

alisman commented on src/shared/components/enhancedReactTable/EnhancedReactTable.tsx:
```diff
@@ -165,6 +217,29 @@ export default class EnhancedReactTable<T> extends React.Component<IEnhancedReac
     }
 
+    componentWillUpdate(nextProps:IEnhancedReactTableProps<T>, nextState:IEnhancedReactTableState) {
+        if (nextState.filter.length === 0) {
```
probably deserves a comment

adamabeshouse replied:
done

Commit:
trying to use reselect for mergedmutations but having module import problems

Commit:
use casting instead of Object assign for setState

Commit:
add comment in componentWillUpdate
